runtime/testy/Ansible.py

```diff
@@ -222,16 +228,101 @@ def start(self):
+
+        stateQueue.put([SM_COMMANDS.SEND_ADDR, [PROCESS_NAMES.TCP_PROCESS]])
+        self.dawn_ip = pipe.recv()[0]
+
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sock.connect((self.dawn_ip, TCPClass.PORT))
+
+    def sender(self, badThingsQueue, stateQueue, pipe):
+        def packageMessage(data):
+            try:
+                proto_message = notification_pb2.Notification()
+                proto_message.header = notification_pb2.Notification.CONSOLE_LOGGING
+                proto_message.console_output = data
+                return proto_message.SerializeToString()
+            except Exception as e:
+                badThingsQueue.put(BadThing(sys.exc_info(), 
+                    "TCP packager crashed with error: " + str(e),
+                    event = BAD_EVENTS.TCP_ERROR,
+                    printStackTrace = True))
+        def packageConfirm(confirm):
+            try:
+                proto_message = notification_pb2.Notification()
+                if confirm:
+                    proto_message.header = notification_pb2.Notification.STUDENT_RECEIVED
+                else:
+                    proto_message.header = notification_pb2.Notification.STUDENT_NOT_RECEIVED
+                return proto_message.SerializeToString()
+            except Exception as e:
+                badThingsQueue.put(BadThing(sys.exc_info(), 
+                    "TCP packager crashed with error: " + str(e),
+                    event = BAD_EVENTS.TCP_ERROR,
+                    printStackTrace = True))
+        while True:
+            try:
+                rawMessage = pipe.recv()
+                nextCall = time.time()
+                nextCall += 1.0/TCPClass.TCP_HZ
+                data = rawMessage[1]
+                if rawMessage[0] == ANSIBLE_COMMANDS.STUDENT_UPLOAD:
+                    packedMsg = packageConfirm(data)
+                elif rawMessage[0] == ANSIBLE_COMMANDS.CONSOLE:
+                    packedMsg = packageMessage(data)
+                else:
+                    continue
+                if packedMsg is not None:
+                    self.sock.sendall(packedMsg)
+                #Sleep for throttling thread
+                time.sleep(max(nextCall - time.time(), 0))
+            except Exception as e:
+                badThingsQueue.put(BadThing(sys.exc_info(), 
+                    "TCP sender crashed with error: " + str(e),
+                    event = BAD_EVENTS.TCP_ERROR,
+                    printStackTrace = True))
+
+    def receiver(self, badThingsQueue, stateQueue, pipe):
+        def unpackage(data):
+            received_proto = notification_pb2.Notification()
+            received_proto.ParseFromString(data)
+            return received_proto
+        try:
+            while True:
+                recv_data, addr = self.sock.recvfrom(2048)
+                unpackagedData = unpackage(recv_data)
+                if not unpackagedData:
```

Comment:
Based on this statement, `recvfrom` appears to be non-blocking. In that case, this loop is currently busy-waiting. You should probably add a blocking call to `select`, or set the socket to blocking mode.